### DISCUSSION TITLE
I don't know what kind of php extensions can be developed using zephir？

### DISCUSSION THREAD

**Discussion by AuroraYolo** [zephir-lang/zephir#2412] [Ideas]
Can you give an example?

**Comment by Jeckerson** [comment]
Any type of logic that you want to ship as single compiled PHP extension. Any composer package can be rewritten in Zephir and shipped as PHP extension...

**Reply by trane77** [reply]
Hi @Jeckerson, I've just found Zephir lang, it seems interesting, but I'm trying to figure out any situation where Zephir would be a better choice than a regular php library.

Any existing project, tutorial or example would really help to figure this out. 
Thank you

**Reply by rernesto** [reply]
> Any existing project, tutorial or example would really help to figure this out. Thank you

Hi @trane77 this is a work in progress but could be useful for you and gives you some ideas. 

https://github.com/coral-media/php-ext

I have been dissecting Phalcon to understand some stuff poorly or not documented at all. Thats other project you could look at.

**Comment by dawid-czarnecki** [comment]
I would like to write an extension that will log access to certain PHP variable arrays or superglobals. Example would be:
```php
<?php
print($_SERVER['REMOTE_ADDR']);
?>
```

In that case when the request is being made the extension would log to a file `REMOTE_ADDR of $_SERVER variable accessed`. Ideally with a file name and a line number.

Is it possible to do using Zephir?

**Reply by Jeckerson** [reply]
Yes, but you will can your custom logic, not native PHP functions. You can try to implement it via PHP's hooks - https://docs.zephir-lang.com/0.12/en/lifecycle

Or use another extension, such as this example - https://www.php.net/manual/en/function.uopz-set-hook.php

**Reply by dawid-czarnecki** [reply]
I just revisited this topic and realized that uopz-set-hook won't do the job since I don't want to hook function print but any access to my super global. Here are a few other examples:
```php
$x = base64_encode($_SERVER["HTTP_ACCEPT_LANGUAGE"]) // 1
$y = $_GET["param1"] //2
$z = $y["sub_param"] // 3
```
@Jeckerson Do you think it's possible with Zefir?

**Comment by tamr** [comment]
For security purposes it would be great to have the extension to log any outgoing requests (get, post.. url, payload, headers) of file_get_contents, fopen, curl, fsockopen, mail, etc.

**Comment by rernesto** [comment]
https://github.com/coral-media/php-ext

Thats my new baby, so far I am using Zephir to wrap around C libraries. I am still developing primitives, but I could not find a Bayes implementation in C meeting my expectations and requirements, so I probably will end up implementing the Bayes (Multinomial, Gaussian, Bernoulli, etc) myself in Zephir.

I see Zephir as a good tool to bring Scientific Computing to PHP, even without relying in native C libraries (like I am doing and will be optimal for most scenarios), writing the implementations on Zephir should be better than pure PHP. There is the option of using FFI, but for Scientific Computing algorithm could be a nightmare when comes to memory handling.
